L1Trigger/Phase2L1ParticleFlow/src/CaloClusterer.cc

```diff
@@ -0,0 +1,640 @@
+          continue;
+        }
+      }
+      if (cluster.et > (photon ? minPhotonEt_ : minHadronEt_)) {
+        ret->emplace_back(cluster.et,
+                          cluster.eta,
+                          cluster.phi,
+                          cluster.ecal_et > 0 ? std::max(cluster.et - cluster.ecal_et, 0.f) / cluster.ecal_et : -1,
+                          photon);
+        if (setRefs) {
+          for (auto &pair : cluster.constituents) {
+            assert(pair.first != 0);
+            if (pair.first > 0) {  // 1+hcal index
+              ret->back().addConstituent(edm::Ptr<l1t::PFCluster>(hcal, +pair.first - 1), pair.second);
+            } else {  // -1-ecal index
+              ret->back().addConstituent(edm::Ptr<l1t::PFCluster>(ecal, -pair.first + 1), pair.second);
+            }
+          }
+        }
+      }
+    }
+  }
+  return ret;
+}
+
+l1tpf_calo::SimpleCaloLinker::SimpleCaloLinker(const edm::ParameterSet &pset,
+                                               const SingleCaloClusterer &ecal,
+                                               const SingleCaloClusterer &hcal)
+    : SimpleCaloLinkerBase(pset, ecal, hcal), ecalToHCal_(*grid_) {}
+
+l1tpf_calo::SimpleCaloLinker::~SimpleCaloLinker() {}
+
+void l1tpf_calo::SimpleCaloLinker::clear() {
+  clearBase();
+  ecalToHCal_.clear();
+}
+
+void l1tpf_calo::SimpleCaloLinker::run() {
+  unsigned int i, ncells = grid_->size();
+
+  const EtGrid &hraw = hcal_.raw();
+  const IndexGrid &ecals = ecal_.indexGrid();
+  const IndexGrid &hcals = hcal_.indexGrid();
+
+  // for each ECal cluster, get the corresponding HCal cluster and the sum of the neighbour HCal clusters
+  ecalToHCal_.clear();
+  for (i = 0; i < ncells; ++i) {
+    if (ecals[i] >= 0) {
+      if (hcals[i] >= 0) {
+        ecalToHCal_[i].ptLocalMax = hcal_.cluster(i).et;
+      } else {
+        float tot = 0;
+        for (int ineigh = 0; ineigh < 8; ++ineigh) {
+          tot += hcal_.cluster(grid_->neighbour(i, ineigh)).et;
+        }
+        ecalToHCal_[i].ptOverNeighLocalMaxSum = tot ? ecal_.cluster(i).et / tot : 0;
+      }
+    }
+  }
+
+  clusterIndex_.fill(-1);
+  clusters_.clear();
+  CombinedCluster cluster;
+  // promote HCal clusters to final clusters
+  for (i = 0; i < ncells; ++i) {
+    if (hcals[i] >= 0) {
+      const Cluster &hcal = hcal_.cluster(i);
+      cluster.clear();
+      cluster.constituents.emplace_back(+i + 1, 1);
+      if (ecalToHCal_[i].ptLocalMax > 0) {
+        // direct linking is easy
+        const Cluster &ecal = ecal_.cluster(i);
+        if (ecal.et + hcal.et > minHadronRawEt_) {
+          cluster.ecal_et = ecal.et;
+          cluster.hcal_et = hcal.et;
+          cluster.et = cluster.ecal_et + cluster.hcal_et;
+          float wecal = cluster.ecal_et / cluster.et, whcal = 1.0 - wecal;
+          cluster.eta = ecal.eta * wecal + hcal.eta * whcal;
+          cluster.phi = ecal.phi * wecal + hcal.phi * whcal;
+          // wrap around phi
+          if (cluster.phi > M_PI)
```

Comment:
use deltaPhi.h

Reply:
is `reco::reduceRange()` not less efficient, when one knows that the input is already in the range [ -2π, 2π] ?

Reply:
used `reco::reduceRange`